idaes/core/util/model_diagnostics.py

```diff
@@ -4122,3 +4264,277 @@ def _collect_model_statistics(model):
     )
 
     return stats
+
+
+class ConstraintTermAnalysisVisitor(EXPR.StreamBasedExpressionVisitor):
+    """
+    Expression walker for checking Constraints for problematic terms.
+
+    This walker will walk the expression and look for summation terms
+    with mismatched magnitudes or potential cancellations.
+
+    Args:
+        term_mismatch_tolerance - tolerance to use when determining mismatched
+            terms
+        term_cancellation_tolerance - tolerance to use when identifying
+            possible cancellation of terms
+
+    Returns:
+        list of values for top-level summation terms
+        list of terms with mismatched magnitudes
+        list of terms with potential cancellations
+        bool indicating whether expression is a constant
+    """
+
+    def __init__(
+        self,
+        term_mismatch_tolerance: float = 1e6,
+        term_cancellation_tolerance: float = 1e-4,
+    ):
+        super().__init__()
+
+        # Tolerance attributes
+        self._log_mm_tol = log10(term_mismatch_tolerance)
+        self._sum_tol = term_cancellation_tolerance
+
+        # Placeholders for collecting results
+        self.canceling_terms = ComponentSet()
+        self.mismatched_terms = ComponentSet()
+
+    def _get_value_for_sum_subexpression(self, child_data):
+        return sum(i for i in child_data[0])
+
+    def _sum_combinations(self, values_list):
+        sums = []
+        for i in chain.from_iterable(
+            combinations(values_list, r) for r in range(2, len(values_list) + 1)
```

Comment:
2 is a magic number! Why 2?

Reply:
We are looking for any combination of terms which cancel, thus the minimum number of terms to consider is 2 (a single term cannot cancel with itself). I can add a comment.

Reply:
What this block of code does is go through combinations of terms in an expression first in groups of 2, then in groups of 3, all the way up to groups of `len(values_list)`. So if we have the sum expression `a+b+c+d`, we first iterate through `(a, b), (a, c), (a, d), (b, c), (b, d), (c, d)`, then iterate through `(a, b, c,), (a, b, d), (a, c, d), (b, c, d)`, then `(a,b,c,d)`.

However, the number of terms you're checking grows exponentially in expression size. In particular, if `len(values_list) == m`, you'll be checking `2 ** m - m -1` terms. I expect that this sort of check will take an extremely long time on any model with `Expressions` of any significant length, much less an extreme chonker like eNRTL.